File: unix-tools/js/uniq.js
import fs from "node:fs"
import readLine from "node:readline";

export async function uniq(files) {
    let counts = new Map();
    if (files.length == 0) {
        await countLines(process.stdin, counts);
    } else {
        for (let f of files) {
            try {
                const stream = fs.createReadStream(f);
                await countLines(stream, counts);
                stream.close();
            } catch (err) {
                console.error("uniq-js", err);
            }
        }
    }

    for (let [line, n] of counts) {
        if (n > 1) {
            console.log(`${n}\t${line}`);
        }
    }
}

async function countLines(stream, counts) {
    const rl = readLine.createInterface({
        input: stream,
    })

    for await (const line of rl) {
        counts.set(line, (counts.get(line) || 0) + 1);
    }
}

File: unix-tools/python/uniq.py
import sys
from typing import Dict, TextIO


def uniq(files: list[str]):
    counts = {}
    if not files:
        countLines(sys.stdin, counts)
    else:
        for arg in files:
            try:
                with open(arg, "r") as f:
                    countLines(f, counts)
            except IOError as err:
                print(f"uniq-py: {err}", file=sys.stderr)

        for line, n in counts.items():
            if n > 1:
                print(f"{n}\t{line}")


def countLines(f: TextIO, counts: Dict[str, int]):
    for line in f:
        line = line.strip()
        counts[line] = counts.get(line, 0) + 1
